### PR TITLE
expand extension actuator interface with migrate and restore

### PR DESCRIPTION
**Pull request by vlvasilev**
**What this PR does / why we need it**:
With this PR we extend the Extensions actuator interface with Migrate and Restore functionality.
 
**Which issue(s) this PR fixes**:
Part of [# 1631](https://github.com/gardener/gardener/issues/1631) 

**Special notes for your reviewer**:

**Release note**:

```improvement developer
Extend the Extensions actuator interface with Migrate and Restore
```

### CODE REVIEW

File: extensions/pkg/controller/extension/reconciler.go [hunk 1]
```diff
@@ -173,39 +174,43 @@ func (r *reconciler) Reconcile(request reconcile.Request) (reconcile.Result, err
...
+	case ex.Annotations[v1beta1constants.GardenerOperation] == v1beta1constants.GardenerOperationRestore:
+		if result, err = r.restore(r.ctx, ex, operationType); err != nil {
+			return result, err
+		}
+		return reconcile.Result{Requeue: r.resync != 0, RequeueAfter: r.resync}, nil
+	default:
+		if result, err = r.reconcile(r.ctx, ex, operationType); err != nil {
+			return result, err
+		}
+		return reconcile.Result{Requeue: r.resync != 0, RequeueAfter: r.resync}, nil
```
Comment (rfranzke) [inline extensions/pkg/controller/extension/reconciler.go]: Hm, why do we have this special handing with the error here? I don't see this for the infrastructure controller, for example: https://github.com/gardener/gardener/blob/master/extensions/pkg/controller/infrastructure/reconciler.go#L104-L115

Can you explain?

Comment (vlvasilev) [reply extensions/pkg/controller/extension/reconciler.go]: Monkey see monkey do.
This error handling was there, before the switch.

Comment (plkokanov) [reply extensions/pkg/controller/extension/reconciler.go]: Hmm I guess we should keep it then? Otherwise looks good.

Comment (rfranzke) [reply extensions/pkg/controller/extension/reconciler.go]: cc @timuthy - can you give an opinion? I think you were contributing it initially (hope I remember correctly)

**Comment by rfranzke** [comment]
/kind/enhancement

File: extensions/pkg/controller/extension/reconciler.go (hunk 1 above)
Comment (timuthy) [reply extensions/pkg/controller/extension/reconciler.go]: Are you specifically asking for the re-queuing? The motivation is that an extension can make use of an automatic re-queuing to allow clean-up jobs, for example. But I don't see a necessity for doing this in the restore case.

Comment (rfranzke) [reply extensions/pkg/controller/extension/reconciler.go]: OK, thanks. Yes, agreed, for restore it's probably not needed. @vlvasilev can you adapt accordingly?

**Review by swilen-iwanow — APPROVED** [review]
lgtm

File: extensions/pkg/controller/extension/reconciler.go (hunk 1 above)
Comment (vlvasilev) [reply extensions/pkg/controller/extension/reconciler.go]: Done

**Review by rfranzke — APPROVED** [review]

**Comment by rfranzke** [comment]
/lgtm
/reviewed/do-not-merge
until g/g@v1.4.0 is released, cc @ialidzhikov

**Pull request merged**